torch/_functorch/functional_assertions_helper.py

```diff
@@ -0,0 +1,103 @@
+from collections import defaultdict
+from typing import Callable, Dict
+
+import torch
+import torch._decomp as decomp
+from torch._ops import OpOverload
+import torch._functorch.config as config
+
+aten = torch.ops.aten
+_decompositions: Dict[str, Dict[OpOverload, Callable]] = defaultdict(dict)
+
+
+class DepTokenStateTracker:
+    """
+    A singleton / global context manager which is used to:
+    - Record the global flag that assertion ops functionalization is enabled
+      during compilation if the context is entered.
+    - Record most recent token which is propagated through functional assertion
+      ops.
+    """
+
+    _dep_token = None
+    _entered: bool = False
+
+    def __enter__(self) -> "DepTokenStateTracker":
+        assert (
+            not DepTokenStateTracker._entered
+        ), "Cannot reenter singleton context `DepTokenStateTracker`."
+        self.reset()
+        DepTokenStateTracker._entered = True
+        return self
+
+    def __exit__(self, exc_type, exc_cal, exc_tb) -> None:
+        assert DepTokenStateTracker._entered, (
+            "Cannot exit singleton context `DepTokenStateTracker` which is not "
+            "entered previously."
+        )
+        self.reset()
+        DepTokenStateTracker._entered = False
+        return None
+
+    @classmethod
+    def reset(cls) -> None:
+        cls._dep_token = None
+
+    @classmethod
+    def set_dep_token(cls, dep_token) -> None:
+        assert cls._entered
+        cls._dep_token = dep_token
+
+    @classmethod
+    def get_dep_token(cls):
+        if cls._dep_token is None:
+            cls._dep_token = aten._make_dep_token()
+
+        return cls._dep_token
+
+
+class FunctionalAssertionsHelper:
+    @staticmethod
+    def can_functionalize(needs_autograd: bool) -> bool:
```

Comment:
dumb nit: rename to `can_functionalize_asserts()` just to make it clear? we have many flavors of functionalization now 😛